pkg/tbtc/node.go

```diff
@@ -205,3 +265,203 @@ func (n *node) joinDKGIfEligible(seed *big.Int, startBlockNumber uint64) {
+	chainConfig *ChainConfig,
+) *dkgRetryLoop {
+	// Pre-compute the 8-byte seed that may be needed for the random
+	// retry algorithm. Since the original DKG seed passed as parameter
+	// can have a variable length, it is safer to take the first 8 bytes
+	// of sha256(seed) as the randomRetrySeed.
+	seedSha256 := sha256.Sum256(seed.Bytes())
+	randomRetrySeed := int64(binary.BigEndian.Uint64(seedSha256[:8]))
+
+	return &dkgRetryLoop{
+		memberIndex:              memberIndex,
+		selectedOperators:        selectedOperators,
+		inactiveOperatorsSet:     make(map[chain.Address]bool),
+		chainConfig:              chainConfig,
+		attemptCounter:           0,
+		attemptStartBlock:        initialStartBlock,
+		randomRetryCounter:       0,
+		randomRetrySeed:          randomRetrySeed,
+		delayBlocks:              5,
+		delayBlocksBumpFrequency: 100,
+		delayBlocksBumpFactor:    20,
+	}
+}
+
+// dkgAttemptParams represents parameters of a DKG attempt.
+type dkgAttemptParams struct {
+	index           uint
+	startBlock      uint64
+	excludedMembers []group.MemberIndex
+}
+
+// dkgAttemptFn represents a function performing a DKG attempt.
+type dkgAttemptFn func(*dkgAttemptParams) (*dkg.Result, error)
+
+// start begins the DKG retry loop using the given DKG attempt function.
+// The retry loop terminates when the DKG result is produced or the ctx
+// parameter is done, whatever comes first.
+func (drl *dkgRetryLoop) start(
+	ctx context.Context,
+	dkgAttemptFn dkgAttemptFn,
+) (*dkg.Result, error) {
+	// All selected operators should be qualified for the first attempt.
+	qualifiedOperatorsSet := drl.selectedOperators.Set()
+
+	for {
+		drl.attemptCounter++
+
+		// Check the loop stop signal.
+		if ctx.Err() != nil {
+			return nil, fmt.Errorf(
+				"dkg retry loop received stop signal on attempt [%v]",
+				drl.attemptCounter,
+			)
+		}
+
+		// In order to start attempts >1 in the right place, we need to
+		// determine how many blocks were taken by previous attempts. We assume
+		// the worst case that each attempt failed at the end of the DKG
+		// protocol. That said, we need to increment the previous attempt start
+		// block by the number of blocks equal to the protocol duration and
+		// by some additional delay blocks. We need a small fixed delay in
+		// order to mitigate all corner cases where the actual attempt duration
+		// was slightly longer than the expected duration determined by the
+		// dkg.ProtocolBlocks function. For example, the attempt may fail at
+		// the end of the protocol but the error is returned after some time
+		// and more blocks than expected are mined in the meantime.
+		// Additionally, we want to strongly extend the delay period
+		// periodically in order to give some additional time for nodes to
+		// recover and re-fill their internal TSS pre-parameters pools.
+		if drl.attemptCounter > 1 {
+			delayBlocks := drl.delayBlocks
+			if drl.attemptCounter%drl.delayBlocksBumpFrequency == 0 {
```

Comment:
```suggestion
			if drl.attemptCounter % drl.delayBlocksBumpFrequency == 0 {
```

Reply:
This is automatically formatted by `gofmt`